### PR TITLE
Yusen/Add BM25 and TF-IDF for Query-based Summarization

### PR DESCRIPTION
**Pull request by chatc**
TODO: 
1. Modify dependency.txt
2. Change the Static var is_neural

### CODE REVIEW

**Comment by chatc** [comment]
All tasks are done. Ready for review now.

**Review by niansong1996 — COMMENTED** [review]
Sorry that I forgot to submit my reviews... Can you respond to them? thanks.

File: model/query_based/base_query_based_model.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                 max_output_length: int = None,
+                 model_backend: SummModel = TextRankModel, **kwargs
+                 ):
+        super(QueryBasedSummModel, self).__init__(trained_domain = trained_domain, max_input_length = max_input_length, max_output_length = max_output_length)
+        self.model = model_backend(**kwargs)
+
+    def _retrieve(self, instance: List[str], query: List[str], n_best) -> List[str]:
+        raise NotImplementedError()
+
+    def summarize(self, corpus: Union[List[str], List[List[str]]], queries: List[str] = None,
```
[the remark this answers is not in the document]
Comment (chatc) [reply model/query_based/base_query_based_model.py]: I think this is one function of the base class: the base class with check the type and raise errors. Query = None indicates that the query is invalid and the further processing with be executed.

File: model/query_based/base_query_based_model.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                  retrieval_ratio: float = 0.5, preprocess=True) -> List[str]:
```
[the remark this answers is not in the document]
Comment (chatc) [reply model/query_based/base_query_based_model.py]: Done!

File: model/query_based/base_query_based_model.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        self.assert_summ_input_type(corpus, queries)
+
+        retrieval_output = [] # List[str]
+        for instance, query in zip(corpus, queries):
```
[the remark this answers is not in the document]
Comment (chatc) [reply model/query_based/base_query_based_model.py:35]: This checks if the corpus and queries are in the correct form. this checking function can prevent the user from inputting some illegal input to the summarizer.

File: model/query_based/base_query_based_model.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                        "Weaknesses: \n - Heavily depends on the performance of both retriever and summarizer.\n")
+        print(f"{basic_description}\n{'#' * 20}\n{more_details}")
+
+
+class Preprocessor:
+    def __init__(self):
+        self.sw = stopwords.words('english')
+        self.stemmer = PorterStemmer()
+
+    def preprocess(self, corpus: List[str], remove_stopwords=True, lower_case=True, stem=False) -> List[str]:
```
[the remark this answers is not in the document]
Comment (chatc) [reply model/query_based/base_query_based_model.py]: Yes, this is for boosting the performance of all retrievers.

**Comment by niansong1996** [comment]
Can you fix the conflict? And what's the status of this branch? Ready to be merged yet?

**Comment by chatc** [comment]
> Can you fix the conflict? And what's the status of this branch? Ready to be merged yet?

The conflicts have been solved and the branch is ready to be merged!

**Comment by niansong1996** [comment]
Awesome! Thx

**Pull request merged**